sklearn/preprocessing/_discretization.py

```diff
@@ -231,32 +246,72 @@ def fit(self, X, y=None, sample_weight=None):
 
         n_samples, n_features = X.shape
 
-        if sample_weight is not None and self.strategy == "uniform":
-            raise ValueError(
-                "`sample_weight` was provided but it cannot be "
-                "used with strategy='uniform'. Got strategy="
-                f"{self.strategy!r} instead."
-            )
-
         if self.subsample is not None and n_samples > self.subsample:
             # Take a subsample of `X`
+            # When resampling, it is important to subsample **with replacement** to
+            # preserve the distribution, in particular in the presence of a few data
+            # points with large weights.
```

Comment:
Could you link me to the discussion or notebook that lead to this conclusion? I had a quick scroll through the PR discussion but there are a lot of comments, so I didn't immediately find it.

I don't know if with or without replacement is the "right thing" to do, so I'd be interested to see how you figure this out. A/The only sampling method I know that uses resampling is https://en.wikipedia.org/wiki/Bootstrapping_(statistics) but I am not sure if this is what we are doing here. I think we are downsampling which is different. TL;DR: I'm interested to learn more :D

---

One thing I remember is that having a large spread in weights (say some samples have a weight of around 1 and then other samples have a weight of 100) leads to some parts of the distribution being known with much less precision. A bit like having five samples with a weight of 10 vs fifty samples with a weight of 1. In the latter case you will have a much better estimate of the shape of the distribution that your samples are from. But I don't know how this plays into what downsampling strategy to use.

Reply:
In this https://github.com/scikit-learn/scikit-learn/pull/29907#issuecomment-2407740996 I posted the results of some experiments with the different strategies, that I ran using this notebook https://gist.github.com/snath-xoc/ef5dbe53ca4f02655fe99173e9ab1198

Reply:
I agree that it would be good to link to a good reference to back this claim from the comment in the source code.

We could link to https://github.com/scikit-learn/scikit-learn/pull/29907#issuecomment-2407740996, although it does not really focus on weighted sampling with vs without replacement, but instead on [uniform sampling followed by weighted quantile computation] vs [weighted sampling followed by regular quantile computation], which is only indirectly related to the topic of the example.

Instead, we could point to the new test case:

```suggestion
            # points with large weights. You can check this by setting `replace=False`
            # in sklearn.utils.test.test_indexing.test_resample_weighted and check that
            # it fails as a justification for this claim.
```